### PR TITLE
Proposed behavior change for DataLoader with IterableDataset

### LINKED ISSUE #40972: len of dataloader when using iterable dataset does not reflect batch size
## 🐛 Bug

If I construct an iterable dataset of length 100, and want to then use that in a dataloader with batch_size=4, one would expect the length of the dataloader to be 25, but instead it is 100. This is because it is implemented as
```
        if self._dataset_kind == _DatasetKind.Iterable:
            length = self._IterableDataset_len_called = len(self.dataset)
            return length
```

## To Reproduce

```
from torch.utils.data import DataLoader, IterableDataset, Dataset
test_items = list(range(100))
class TestDataset(Dataset):
    
    def __init__(self, test_items):
        self.x = test_items

    def __getitem__(self, item):
        return self.x[item]

    def __len__(self) -> int:
        return len(self.x)

class TestIterableDataset(IterableDataset):
    
    def __init__(self, test_items):
        self.x = test_items

    def __iter__(self):
        return iter(self.x)

    def __len__(self) -> int:
        return len(self.x)

print(len(TestIterableDataset(test_items)))
print(len(TestDataset(test_items)))
print(len(DataLoader(TestIterableDataset(test_items), batch_size=4)))
print(len(DataLoader(TestDataset(test_items), batch_size=4)))
```
this prints:
```
100
100
100
25
```

## Expected behavior

I would expect this to print:
```
100
100
25
25
```
## Environment

Please copy and paste the output from our
torch '1.5.1'

## Additional context

If people agree this should be fixed, I would love to put out a PR for it! Seems manageable and I've been wanting to contribute for a while now :)

### PR DESCRIPTION
**Pull request by jlucier**
See recent comments on #38925 for more context, but I'll reproduce a synthesis here.

## Background
While well intentioned, the `__len__` function of `DataLoader` is inaccurate when used with multi-process data loading and data is being sharded across workers. This is due to the fact that there is no way to know exactly how data is being sharded amongst workers (evenly, or slightly unevenly) and thus one cannot know how `batch_size` will divide into the number of examples for which each worker ends up being responsible. This results is an often mis-calculated number of batches that are expected from the `DataLoader`, and this often ends up producing warnings [here](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L353).

It is *probably* fine for `__len__` to provide a coarse estimate of the number of batches that will be yielded from the loader, but I suppose it's hard to know if that is a brittle assumption. However, this function plays a roll in the warnings the `DataLoader` produces if an `IterableDataset` produces more data than expected based on its length because it is setting `self._IterableDataset_len_called`.

## Proposal
This proposed modification changes how this potential bug is monitored by instead looking at the *number of samples* yielded by iterating over the `DataLoader` rather than the *number of batches*. This is helpful because regardless of how the user chooses to shard data amongst workers, how the data aligns into batches, or whether `drop_last` is on or off, the number of samples returned by the loader can be properly monitored and compared with `len(dataset)` in order to identify potential bugs with poorly sharded data.

The plan (as revised by @SsnL) is to have the the `_BaseDatasetFetcher` (and subclasses) return the number of samples in addition to the fetched data for each call to `fetch`. This `num_samples` is propagated up from the "fetcher" back to the `DataLoader`, which -- like before -- will track the total number of samples it has returned. However, now it can do so by accumulating the `num_samples` the fetcher (and/or workers) are giving it rather than assuming it understands exactly the right number of samples that have been returned.

## Discussion
The main advantage of this change is that the warning (alerting users to potential duplicated data and sharding issues when using `IterableDataset`) can be accurately monitored and provide helpful feedback to the user regardless of how they've implemented sharding of their data.

cc @SsnL 

(P.S. #40281 is at odds with this change and the two should be considered mutually exclusive, that PR does not fix these issues)

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -352,7 +352,11 @@ def _next_data(self):
 
     def __next__(self):
         data = self._next_data()
-        self._num_yielded += 1
+        # num_yielded and the value of _IterableDataset_len_called are in units of "samples" rather than
+        # batches, so when using auto_collation, we need to increment by the length of the batch we're
+        # yielding
+        self._num_yielded += len(data) if self._auto_collation else 1
```
Comment (SsnL) [inline torch/utils/data/dataloader.py:358]: this is wrong through, because of collation. so what we can do is to gather the total #yields from workers and then also return them, maybe from `_next_data`.

**Review by ssnl — COMMENTED** [review]
would also need a test

File: torch/utils/data/dataloader.py [hunk 2]
```diff
@@ -352,7 +352,11 @@ def _next_data(self):
...
+
```
Comment (SsnL) [inline torch/utils/data/dataloader.py:359]: also worth adding a comment that things like `drop_last` will not be falsely triggering the following check.

File: torch/utils/data/dataloader.py (hunk 1 above)
Comment (jlucier) [reply torch/utils/data/dataloader.py:358]: Is this wrong because data often something like a tuple or dict rather than a Tensor or custom object that can implement `__len__`?

Comment (SsnL) [reply torch/utils/data/dataloader.py:358]: It is wrong because `data` will be the collated version. For example, if the data is of the form `(tensor_image, int_label)`, a collated batch of 10 samples will be a 2-tuple, `data = (batched_tensor_images, batched_tensor_labels)`, with each element having a leading dimension (batch dimension) of size 10. Yet `len(data) = 2`. Moreover, in general one can not detect number of samples from yielded data because of possible custom collate function. So this would require some deeper modification of the communication logic, and specifically including the workers communicating how many `next` they have called.

**Comment by jlucier** [comment]
Sorry for the delay, I'm planning on coming back to this and implementing what @SsnL suggested. Tracking the number of `yield` calls from the workers in order to properly check for more data being yielded than should be.

**Comment by SsnL** [comment]
@jlucier If you don't have time to implement this, I'd be happy to do it and push to your fork, if you want me to.

**Comment by jlucier** [comment]
@SsnL That would be awesome!

This dropped down lower in my TODOs because we have a workaround in place for existing releases (e.g. 1.5.1).

**Comment by turian** [comment]
@SsnL Do you know when this will be released? Thank you.

**Comment by SsnL** [comment]
Need a review from @VitalyFedyunin
